### ISSUE TITLE
Big response time for findbyid endpoint

### ISSUE THREAD

**Issue by tadjik1** [pelias/placeholder#102]
Hey guys,
we are using placeholder service only for localisation, so that we are consuming only `findbyid` endpoint and we've noticed that it's slow for the moment (~30ms).
Using debugger I found that most problematic place is `JSON.parse` function.

![screen shot 2018-05-28 at 11 32 02](https://user-images.githubusercontent.com/1452895/40610422-977ad8f6-6272-11e8-86cc-1c6097bff11b.png)

The reason is structure of table `docs`. It contains 2 columns, id and json. Column `json` has a TEXT type and it contains huge (up to ~8.5 characters). These values then parse with `JSON.parse` in order to get some manipulations. Parsing is blocking synchronous operation and it takes some time for server to respond.

The most naive and natural way to make it faster is to decrease amount of json that should be parsed in these cases.
This could be done in several possible ways:
1. more difficult one is to change DB scheme and disassemble containing document to columns.
2. and easiest way is to use JSON type for this column instead. It allows to perform `json_extract` (and some others) operation on DB level, so that we need to parse/stringify much smaller strings and objects.
Not sure about other endpoints (as I said we are not using them), but for localisation it could be done pretty easy.

SQL example (assuming that we want to get german names for several objects:
```
SELECT id, json_extract(json, '$.names.deu[0]') as name 
FROM docs 
WHERE id IN (102063941,102191581,85682555,85633111)
```

And final results of load test (I was using ab):
```
ab -n 1000 -c 100  "http://localhost:3002/parser/findbyid?ids=102063941,102191581,85682555,85633111,102064231,85682523,102063845,102063889,85687045,101807185,404473841,85633337,85682447,85633105,102064007,101875803,102064083,85688637,102081727,85927929,102191575,85633793"

Percentage of the requests served within a certain time (ms)
  50%    325
  66%    328
  75%    330
  80%    334
  90%    338
  95%    339
  98%    351
  99%    357
 100%    362 (longest request)
```

```
ab -n 1000 -c 100  "http://localhost:3002/parser/translations?ids=102063941,102191581,85682555,85633111,102064231,85682523,102063845,102063889,85687045,101807185,404473841,85633337,85682447,85633105,102064007,101875803,102064083,85688637,102081727,85927929,102191575,85633793&lang=eng"

Percentage of the requests served within a certain time (ms)
  50%     83
  66%     86
  75%     88
  80%     90
  90%    104
  95%    114
  98%    124
  99%    125
 100%    127 (longest request)
```

**Comment by missinglink** [comment]
hi @tadjik1 

From memory, I am already stripping out all non-essential JSON properties, so I doubt we can trim the byte size any more without sacrificing functionality :(

To your points:

1. If we normalized the JSON properties, it would still require a JOIN operation in order to assemble the JSON back together again, which probably wouldn't result in much of a performance benefit.

2. Using `json_extract` is pretty easy to do, as you noted, looking at the sqlite src code, the `JSON` type is basically just an alias for `TEXT`, so no changes need to be made to the schema for this to work efficiently:

> ** For the time being, all JSON is stored as pure text.  (We might add
** a JSONB type in the future which stores a binary encoding of JSON in
** a BLOB, but there is no support for JSONB in the current implementation.
** This implementation parses JSON text at 250 MB/s, so it is hard to see
** how JSONB might improve on that.)

I think the issue is that the `findById` function was intended to return **all** the data and let the nodejs code 'pick-and-choose' what names etc it wants to use, which is taking a bunch of time.

There might be people relying on that functionality, so I'm kinda hesitant to change it.

Instead, we could add a new function which returns only a subset of the fields, and call it something different, that way it would maintain backwards compatibility but also allow for an optimization for other use-cases.

I'm guessing the use-case you have in mind is passing the desired language code down to sqlite and having it **only** return the language and properties you requested?

**Comment by missinglink** [comment]
also you wrote `huge (up to ~8.5 characters)` which I assume is a typo? :)

**Comment by missinglink** [comment]
here's a link to the sqlite json1 extension source: https://www.sqlite.org/src/artifact/02f5f0d9c131df79

**Comment by missinglink** [comment]
also, because of the way the [codec is written](https://github.com/pelias/placeholder/blob/master/lib/DocStore.js#L79-L86) you could try another codec, but I'm guessing `JSON.parse` is already pretty optimized in v8

**Comment by missinglink** [comment]
another option we just discussed is to implement a client-side cache, so the results are only parsed once by the `pelias/api` machines and held in RAM.

something like https://github.com/isaacs/node-lru-cache could be used.

this would probably be even more efficient as it would prevent additional network hops for the same data.

**Comment by missinglink** [comment]
@tadjik1 would you be interested in drafting a PR to implement a LRU cache in https://github.com/pelias/api/blob/master/service/configurations/Language.js ?

**Comment by tadjik1** [comment]
@missinglink sorry for my confusing explanation. 
yes, new endpoint that provides only localised names for particular object is that I meant. so instead of getting whole documents and serialise/deserialise them and  we can pick only name in required language. as one parameter (and with regards to string length it was a typo, it's up to ~8.5K characters).
so something like 
`GET /parser/translations?ids=1,2,3&lang=deu` 
which returns 
`{1: {name: 'abs'}, 2: {name: 'def'}, 3: {name: null}}`.

introducing cache on client(s) could be also a case, but I would prefer optimise operations on placeholder side first.

**Comment by orangejulius** [comment]
Ultimately, an LRU cache style implementation would probably offer the best performance, but adding a `lang` param is very easy, avoids the complexities of caching, and is useful even if we do implement caching someday.  It's even already been suggested in https://github.com/pelias/placeholder/issues/42

Let's start with the `lang` param and if that ever becomes not enough, we can start caching.

**Comment by missinglink** [comment]
:+1: I'm happy to accept a PR on this repo to add a new endpoint.

**Comment by missinglink** [comment]
FYI it's possible to have a fallback behaviour for names using sql `IFNULL()`

**Comment by orangejulius** [comment]
How does everyone feel about Placeholder performance after #128?

From our geocode.earth stats, we are showing p99 latencies (including time to parse the JSON on the API side, of around 100ms), which is quite good. Median latency is only 10ms as well. Seems like for now we don't need to do any more work to achieve acceptable performance, and the complexity of caching isn't worth it for now.

<img width="1081" alt="Screen Shot 2019-05-10 at 4 06 41 PM" src="https://user-images.githubusercontent.com/111716/57533341-d083ab00-733d-11e9-9b0d-adcf78b4329c.png">
<img width="1127" alt="Screen Shot 2019-05-10 at 4 06 06 PM" src="https://user-images.githubusercontent.com/111716/57533342-d083ab00-733d-11e9-96e4-20b76fab65bb.png">

**Comment by orangejulius** [comment]
I'm closing this issue, but if anyone notices any other performance concerns, feel free to comment here or open a new issue! :dash:
